Introduce dedup+zstd suppport.

kumpera opened the pull request:
This is done by introducing i_logger_extensions which hides all dedup behavior and state from the rest of the codebase.
The main touchpoint is in logger_facade.cc, that uses it to produce transformed events.
The use of the is_enabled() call is to shortcut to a fast path when available.

Finally, this commit includes test_tools/log_parser/parser.py which can parse rlclientlib output when using the file
backend.

I can further slice this change by taking the i_logger_extensions out if desireable.

ataymano commented on rlclientlib/dedup-internals.h:
```diff
@@ -0,0 +1,108 @@
+    float value() const { return _current; }
+  private:
+    float _current;
+    const float _weigth;
```
[nit] typo

ataymano commented on rlclientlib/dedup-internals.h:
```diff
@@ -0,0 +1,108 @@
+
+  class ewma {
+  public:
+    ewma(float initial = 1, float weigth = 0.5): _current(initial), _weigth(weigth) {}
```
explicit

ataymano replied:
oh nevermind

olgavrou commented on rlclientlib/dedup.cc:
```diff
@@ -1,83 +1,315 @@
+{
+  i_time_provider* _time_provider = nullptr;
+  l::dedup_info_serializer ser;
+  const auto now = _time_provider != nullptr ? _time_provider->gmt_now() : timestamp();
```
how can `_time_provider` not be `nullptr` if it is just set to `nullptr` two lines above?

kumpera replied:
This is a TODO that slipped over, good catch.

olgavrou reviewed (commented):
Some comments:

- there is a mismatch on passing arguments between the PR and the existing code: `type &arg` or `type *arg` vs existing: `type& arg` or `type* arg`
- `dedup-internals.h` rename to `dedup_internals.h` (underscore instead of hyphen)
- there are some comments in the code starting with `XXX` and some of them seem to be resolved or TODOs, can they be removed or made TODOs?
- I like the logger extension logic :)

kumpera commented on rlclientlib/dedup-internals.h:
```diff
@@ -0,0 +1,108 @@
+    int compress(generic_event::payload_buffer_t& input, api_status* status) const;
+    int transform_payload_and_add_objects(const char* payload, std::string& edited_payload, generic_event::object_list_t& object_ids, api_status* status);
+
+    //test helpers, don't use them directly
```
Hmmm, good point. I wonder what the rest of the codebase does, as it this probably isn't the first instance of this.

kumpera commented on rlclientlib/dedup-internals.h:
```diff
@@ -0,0 +1,108 @@
+    std::mutex _mutex;
+  };
+
+  static const char* DEDUP_DICT_EVENT_ID = "3defd95a-0122-4aac-9068-0b9ac30b66d8";
```
So it's visible to the test suite.

olgavrou replied:
ah ok, can it be named `DEDUP_DICT_EVENT_ID_TEST` then?

kumpera commented on unit_test/event_logger_test.cc:
```diff
@@ -0,0 +1,37 @@
+namespace l = reinforcement_learning::logger;
+namespace fb = flatbuffers;
+
+struct my_async_batcher : public l::i_async_batcher<r::generic_event>\
```
Ooops, old file! This was used to test an old design.

kumpera commented on rlclientlib/dedup-internals.h:
```diff
@@ -0,0 +1,108 @@
+    string_view get_object(generic_event::object_id_t aid);
+    float get_ewma_value() const;
+
+    //FIXME pass a value-oblivious interator to the keys
```
I'm unhappy with passing the map. Changed to templated methods and iterators.

kumpera commented:
> Some comments:
> 
>     * there is a mismatch on passing arguments between the PR and the existing code: `type &arg` or `type *arg` vs existing: `type& arg` or `type* arg`

Fixed.
> 
>     * `dedup-internals.h` rename to `dedup_internals.h` (underscore instead of hyphen)

Fixed.

> 
>     * there are some comments in the code starting with `XXX` and some of them seem to be resolved or TODOs, can they be removed or made TODOs?
> 
Fixed

ataymano commented on rlclientlib/dedup.cc:
```diff
@@ -1,83 +1,288 @@
+  return uniform_hash(start, size, 0);
+}
+
+generic_event::object_id_t dedup_dict::add_object(const char*start, size_t length)
```
looks like it can be called concurrently? do we need to introduce guard here?

kumpera replied:
I designed dedup_dict as explicitly thread unsafe.
All synchronization is done on `dedup_state`.

olgavrou commented on rlclientlib/dedup.cc:
```diff
@@ -1,83 +1,288 @@
   }
 
-  int dedup_dict::transform_payload_and_add_objects(const char* payload, std::string& edited_payload, generic_event::object_list_t& object_ids, api_status* status)
+  int finalize(api_status* status)
```
`status` to default to `nullptr`, same as all other method signatures with `status`?

kumpera replied:
This usage of default values on internal APIs is wrong and we must stop doing it.
There is never a reason why we'd internally not pass a `status` objects. It only enables us to make mistakes.

olgavrou commented on rlclientlib/dedup_internals.h:
```diff
@@ -0,0 +1,140 @@
+    const static int ZSTD_DEFAULT_COMPRESSION_LEVEL = 1;
+
+    explicit zstd_compressor(int level);
+    int compress(generic_event::payload_buffer_t& input, api_status* status) const;
```
same here? default `status` to `nullptr`?

kumpera replied:
same as above

olgavrou commented on rlclientlib/dedup_internals.h:
```diff
@@ -0,0 +1,140 @@
+    int compress(generic_event::payload_buffer_t& input, api_status* status) const;
+    int transform_payload_and_add_objects(const char* payload, std::string& edited_payload, generic_event::object_list_t& object_ids, api_status* status);
+
+    i_time_provider* get_time_provider() { return _time_provider.get(); }
```
as far as I can see, we don't check anywhere that time_provider is `nullptr` or not when it is passed into the `_time_provider` unique_ptr so it could be `nullptr` here resulting in a segmentation fault

kumpera replied:
`_time_provider` is a value of type `unique_ptr`. Accessing it will never result in a segfault.
Calling `get()` on it will either return a valid pointer or `nullptr`.

olgavrou commented on rlclientlib/dedup_internals.h:
```diff
@@ -0,0 +1,140 @@
+    int finalize(generic_event& evt, api_status* status);
+  private:
+    dedup_state& _state;
+    int _size_estimate;
```
```suggestion
    size_t _size_estimate;
```

is there a reason for it to be an int?
